File: Java/src/deepspace/SpaceCity.java
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author Mario Megías Mateo and Shao Jie Hu Chen.
 */
public class SpaceCity extends SpaceStation {
    
    // Attributes

    private SpaceStation base;
    private ArrayList<SpaceStation> collaborators;

    // Methods

    // Constructor

    SpaceCity(SpaceStation base, ArrayList<SpaceStation> rest) {
        super(base);
        this.base          = base;
        this.collaborators = new ArrayList<SpaceStation>(rest);
    }

    public ArrayList<SpaceStation> getCollaborators() {
        return new ArrayList<SpaceStation>(collaborators);
    }

    @Override
    public float fire() {
        float totalPower =super.fire();
        for (SpaceStation station : collaborators) {
            totalPower += station.fire();
        }
        return totalPower;
    }

    @Override
    public float protection() {
        float totalProtection = super.protection();
        for (SpaceStation station : collaborators) {
            totalProtection += station.protection();
        }
        return totalProtection;
    }

    @Override
    public Transformation setLoot(Loot loot) {
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }

}

File: Ruby/SpaceCity.rb
#encoding:utf-8

module Deepspace
    require_relative 'SpaceStation.rb'
    require_relative 'Transformation.rb'

    class SpaceCity < SpaceStation
        def initialize(base, rest)
            supplies = SuppliesPackage.new(base.ammoPower, base.fuelUnits, base.shieldPower)
            super(base, supplies)

            @base = base
            @collaborators = Array.new(rest)
        end
        
        attr_reader :collaborators

        def fire
            totalPower = super

            @collaborators.each do |c|
                totalPower += c.fire
            end

            return totalPower
        end
        
        def protection
            totalProtection = super

            @collaborators.each do |c|
                totalProtection += c.protection
            end

            return totalProtection
        end

        # Override
        def setLoot(loot)
            super
            return Transformation::NOTRANSFORM
        end
    end # SpaceCity
end # Deepspace